### PR TITLE
feat(docker): Added Multi-Stage build support for Vectara-Answer

### LINKED ISSUE #92: Enhance the docker/Dockerfile
Currently, the Dockerfile depends upon the external commands like npm install, and npm run build to be able to include the build folder in the image. Instead, it should be a multi staged Dockerfile with build and package stage. The build stage builds the code and the package stage uses the artifacts. Idea  is that the Dockerfile should work on its own. This will make the build and push workflow smaller and easier to follow. Also, if someone wants to build locally from the Dockerfile, then it will work without any external dependencies.

### PR DESCRIPTION
**Pull request by SaptarshiSarkar12**
## Changes Proposed
Fixes #92 
- Added three stages in the Dockerfile.
    - `deps`: Install dependencies
    - `build`: Builds the source code
    - `runner`: Starts the website
- Added `docker-compose.yml` to simplify the command to start the container.
    - Using `docker compose up`, a container with the name `vectara-answer` will start.
    - Using `docker compose down`, the containers along with any ports/networks will be automatically closed.

## Screenshots
- Docker compose commands working successfully
    up
    down
- Live Website running
    website

## Note To Reviewers
@ofermend I have some thoughts on the changes I have made and those necessary.
- I think that some documentation changes are also required now. So, please let me know if I can make those changes along with this PR. I would also like to know the certain areas of the docs that need to be modified.
- I think there can be only one port (`4444`) used for both Docker and local dev. Also, the `package.json` and `server` folder in the docker folder is not required anymore.
- I think the environment variables are not configured in the docker compose file. I need some more information about what those will be (previously set or user needs to set it).

### CODE REVIEW

**Comment by ofermend** [comment]
Thanks @SaptarshiSarkar12 - a few thoughts:
1. Good work on the staging in the Dockerfile. 
2. The reason for using 3000 originally (instead of 4444) was to differentiate both environment, so a local developer can test both. Not super critical I agree but I don't see a reason why to change that. Can you please clarify why you want to align both of these on the same port?
3. Note also that docker/run.sh runs the docker container at the moment, and has a mechanism to load some files as volumes into the container (using "-v") which I think breaks with this change. Can you please suggest a way to update run.sh to make it compatible (perhaps in docker-compose up?)
4. Yes I agree docs need to change as well - especially two areas: (1) guidance on how to start/stop containers (2) how to deploy to Render with this new approach (if it changes).

**Comment by SaptarshiSarkar12** [comment]
@ofermend 
- Thank you for the review :smile:! I loved working on this issue :smiley:.
- I think that testing on both the local dev and docker would not be done simultaneously. So, keeping the same port would not cause any problem. What do you think?
- Yeah, I will add the **environment variables** in the docker compose file. Can you please tell me what they are going to be? I could not understand some portion of the **run.sh** file near the line where env variables are passed to docker.
- I will change the docs just before the merge. I want to make the code changes first :sweat_smile:.

**Comment by ofermend** [comment]
- for the ports: my thinking is about developers (e.g. you and me) where we may make a change and want to test in both direct and docker mode and see that it works the same. Not a super common case I agree. My question is: was your intent in making it the same port just to simplify and avoid confusion? What's the downside of keeping two types of ports?
- For docker/run.sh, note the following:
  * the "--env-file .env" is making sure environment variables are loaded in. It's intentionally NOT part of the docker build since based on the configuration used the .env file is generated (when running prepare_config.py)
  * The "-v $correct_path/queries.json:/usr/src/app/build/queries.json" part loads the queries file into the right place in the Docker container (/usr/src/app/build) as if it's a volume connected to the container. This specific way to do it enables the queries to be updated outside of the container an the app automatically changes the queries without a need to reload.

Adding @irfanazam1 also (original author of the issue) for his perspective and guidance.

**Comment by SaptarshiSarkar12** [comment]
@ofermend 
- Yes, I wanted to simplify things. The only difference in **index.js** file in `docker/server` folder and **index.js** in the project root, is about the ports. So, if one port is used, it will remove confusion. What do you think on this perspective?
- Would it be okay if env variables are loaded from a **.env** file which will look like this :point_down: (an example)
    ```env
    API_KEY=as21bjb3k4354454523jbkjkjvkjvkjv
    ENDPOINT=
    ....
    ```
    Also, **.env** files are directly loaded by docker. So, it will further simplify the process. The user can simply put the values in the respective fields in the env file. So, I think **prepare_config.py** would also not be required henceforth.
- For the volume, I think there can be a file `queries.json` in **docker** directory and docker compose will bind it to the container automatically to the required location. (For instance, you can check [**this docker-compose file**](https://github.com/SaptarshiSarkar12/Drifty/blob/master/docker-compose.yaml) which creates **.drifty** folder and links it to the one inside the container.) Let me know what do you think.

**Comment by cjcenizal** [comment]
> Yes, I wanted to simplify things. The only difference in index.js file in docker/server folder and index.js in the project root, is about the ports. So, if one port is used, it will remove confusion. What do you think on this perspective?

I love the effort to simplify things! To preserve the ability to run the Docker server and dev server simultaneously, how about you update the server script to accept a `port` command line argument? Here's an example using the [commander library](https://www.digitalocean.com/community/tutorials/nodejs-command-line-arguments-node-scripts#using-commander). We could default the port to 3000 (for the Docker server) and update the NPM script to be:

```
"serve": "node server/index.js --port=4444",
```

Instead of:

https://github.com/vectara/vectara-answer/blob/1de9e530d628676c2d7c60d9771fa9b73bea4462/package.json#L56

**Comment by SaptarshiSarkar12** [comment]
@cjcenizal I have found a way to utilise the `docker/server/index.js` file (it uses port 4444). So, I think I can push that change. What do you think @cjcenizal @ofermend?

**Comment by cjcenizal** [comment]
@SaptarshiSarkar12 Feel free to push a change so we can review it! To be honest, I didn't understand what you proposed but seeing some code will help clarify things for me. 😅

**Comment by SaptarshiSarkar12** [comment]
@cjcenizal Sorry, I made a mistake in the port number I mentioned in the previous comment. The port for docker should be 3000 for Docker but I wrote 4444 (which was for local dev).
I have an idea :bulb: --- What if I use `process.env.port` variable in index.js file in server directory? Then, only one index file is required. By default, port will be 4444, if `process.env.port` is not null, then port will be as set in the env variable. I think this will reduce redundancy. What do you think @ofermend @cjcenizal?

**Comment by SaptarshiSarkar12** [comment]
@ofermend @cjcenizal Now, the port for docker is 3000.
image
image

**Comment by SaptarshiSarkar12** [comment]
@ofermend @cjcenizal I have added a way to automatically get which port to use based on the env variable **port** set. The latest commit implements my idea. Let me know what you think.

**Review by cjcenizal — APPROVED** [review]
Tested locally and I'm able to run both the local dev build and the Docker build simultaneously. Nice solution @SaptarshiSarkar12!

I'm not super familiar with Docker, so I'll defer to @ofermend and @irfanazam1 regarding your Docker changes.

> Also, the package.json and server folder in the docker folder is not required anymore.

Do you want to remove these files as part of this PR? If so can you push that change so we can see the full scope of changes you're proposing?

Thanks again for working on this!

**Comment by SaptarshiSarkar12** [comment]
@cjcenizal Thank you for the compliments :smile:! I loved working on this issue.
The package.json file in docker folder is required to install only express deps. The server folder in docker folder is not required anymore - I agree :heavy_check_mark:. I will make that change and push it soon.

**Comment by SaptarshiSarkar12** [comment]
@ofermend I have some questions on the env variables and volumes.
- I think there should be one **queries.json** file in the project root which contains the queries and it will be directly bound to the container. So, if the json file gets updated, real-time change in the website should also be seen. This is what I am planning to do about the volumes. What do you think?
- What is the **.env** described in **run.sh** file? What does it contain?
    image

**Comment by ofermend** [comment]
The project is designed to support multiple configurations, each can have its own config.yaml and queries.json. That's why, as you can see in the run.sh, we first run prepare_config ( (https://github.com/vectara/vectara-answer/blob/1de9e530d628676c2d7c60d9771fa9b73bea4462/docker/run.sh#L9) which merges config.yaml (in the folder specified) and the secrets from secrets.toml file into a local .env. Queries remains the same. Then when the docker container is launched, it's using the .env just created and loads the queries.json from local folder (or could be S3 in cloud deployment)

Thus, let's be careful not to remove this functionality. If you have a good suggestion to simplify, then that would be great, but I think the basic requirements are: (1) support the ability to run multiple project with the same local repo (by directing to a folder with the configuarion) (2) similarly, supporting multiple secrets (currently with secrets.toml and multiple profiles)

Does that make sense?

**Comment by SaptarshiSarkar12** [comment]
Yeah, that makes sense. Then, **prepare_config.py** file must be run before `docker compose up` and there is no change in that. @ofermend How does **run.sh** affect the **.env** file? Does it even affect it?
@ofermend Can you please provide me a sample **.env** file that will be generated by **prepare_config.py** file (so that I can test the env variables within docker container)?

**Comment by ofermend** [comment]
run.sh just runs prepare_config.py which creates (overwrites if exists) the .env file in the local folder. Then the docker run command results in that .env being loaded to the Docker container via the --enf-file argument. 

Can you please use run.sh locally to see how the .env looks like?

**Comment by SaptarshiSarkar12** [comment]
@ofermend Yes, I can do that locally. But, I need to have a **secrets.toml** file. Can you please provide me a sample secrets file?
@ofermend I think that python program can also be run inside docker so that the end user does not need to have python installed - which makes things more simple. Should I also attempt for this objective? What do you think?

**Comment by ofermend** [comment]
@SaptarshiSarkar12 
- Please take a look at secrets.example.toml for how the file should look. it is comprised of various "profiles" each with a name (e.g. default or feynman etc). Under each profile you can put multiple lines of secret variables, and in our case of vectara-answer it includes the Vectara API key as you can see in the example
- Yes, I like the idea of running prepare_config inside the docker. For that to work though we will have to make secrets.toml available to the Docker container in addition to the config.yaml. Perhaps we load them as volumes (just like queries.json) and then some code in the docker creates environment variables from those as needed? Would be great to do it this way, but be aware that "copying" those files into the docker image is not a good idea, as it will result in secrets being exposed with a Docker image. This has to be done in run-time where the secrets are "loaded as a volume" and available to the code inside the container.

**Comment by SaptarshiSarkar12** [comment]
Okay. I'll do it. I will create volumes, no issue in that :+1:.

File: docker/Dockerfile
Comment (mirfanazam76) [inline docker/Dockerfile:15]: These all "build" names are making this line little confusing. I would change FROM deps AS build to builder. It will make it easier to understand.

File: server/index.js
Comment (mirfanazam76) [inline server/index.js:5]: nit: I don't know we are using ?? operator anywhere else but we can refactor this line like
const port = process.env.port ?? 4444

File: docker/Dockerfile
Comment (mirfanazam76) [inline docker/Dockerfile:16]: Can't we do all the build stuff in the builder stage?

File: docker/Dockerfile
Comment (irfanazam1) [inline docker/Dockerfile:15]: IMO these all "build" words are making this line a little confusing. We can change the build stage name to builder to make it more clear.

Comment (SaptarshiSarkar12) [reply docker/Dockerfile:15]: Yeah, it will be better. I'll do that.

File: docker/Dockerfile
Comment (SaptarshiSarkar12) [reply docker/Dockerfile:16]: Yes, we can do that. I'll make the necessary changes and push it. Thank you @irfanazam1 for the review.

**Comment by SaptarshiSarkar12** [comment]
@ofermend I have successfully linked the **.env** file to the docker container. But, I am facing issues in the python program in docker. **.env** file does not show up in my file explorer. I think I can try the **prepare_config** program using java and it should work as I have already one working (You can check [this project of mine](https://github.com/SaptarshiSarkar12/Drifty)). Let me know if it will be useful for you.
Also, a point to mention, if python is also run via docker separately (as .env file cannot be empty when even one of the service is started in docker compose), then, the docker commands get a bit complex.

**Comment by ofermend** [comment]
Thanks @SaptarshiSarkar12 
- No, let's not add another programming language (e.g. Java) - already complex enough with two (JS/Python) and I want to avoid more complexity/dependencies. 
- Maybe I can help with the .env.  When executing "  docker run --platform=linux/amd64 -d -v $correct_path/queries.json:/usr/src/app/build/queries.json --env-file .env --name vanswer -p 127.0.0.1:80:3000/tcp vectara-answer" - note that to my understanding of how Docker works - ".env" is NOT copied over inside the container, rather the individual entries in the file are set in the local environment inside the container. If you want to "load" the file to be visibiel inside the container, perhaps use the "-v" option just like the queries.json?

**Comment by SaptarshiSarkar12** [comment]
@ofermend I was telling the instead of python, we can have java. I was not asking you to add java as another language.
I am not facing any issue with **.env** file. I was telling that we can't have both prepare config and website build altogether. If it would have been java, I could have got some solutions as I primarily work with java :sweat_smile:.

**Comment by SaptarshiSarkar12** [comment]
One thing to note, the **.env** file must be present in the project root before starting up docker compose. So, can I edit the **run.sh** file in accordance with this?

**Comment by SaptarshiSarkar12** [comment]
@ofermend I have successfully made changes in **run.sh** file and it works perfectly. I have one question - Should the docker container run in background? Should it show the logs (only possible if it runs in foreground)? 
If it runs in foreground, then, **open http://localhost** will not work until the container stops.

**Comment by ofermend** [comment]
Got it - thank you. Again I hesitate to add Java into the mix. Why don't we leave the "prepre_config" issue aside and perhaps solve it later. I think your PR is valuable enough even without solving this issue.

**Comment by ofermend** [comment]
I think Docker running in background (as it is now with docker run) is best

**Comment by SaptarshiSarkar12** [comment]
Okay then. I am keeping it as it was. No issue :smile:.

**Comment by SaptarshiSarkar12** [comment]
Okay, I will keep the docker container running in background. No issue. Thank you :smile:!

**Comment by SaptarshiSarkar12** [comment]
@ofermend I have made changes in the README file. But, I couldn't figure out the portions of the HOSTING.md file where changes are required. Can you please guide me by showing the areas of changes required?

**Comment by ofermend** [comment]
Thanks @SaptarshiSarkar12 -
1. I think you are creating this PR against an older version of the main branch. Since then there's been significant updates in terms of the README and other docs. Can you please re-sync your PR with the up-to-date main repo and see how to apply the update against that so we can easily merge?
2. For HOSTING.md - that is using the latest released image from docker-hub. So it's NOT directly using the local image and thus I believe should work as it does not without need for modification. Pls let me know if I'm missing anything here (I assume even though we use docker-compose up/down locally, the image itself can still be uploaded to docker-hub and used in the old way e.g. docker run - see here: https://hub.docker.com/repository/docker/vectara/vectara-answer/general

**Comment by SaptarshiSarkar12** [comment]
@ofermend Yes, I will sync my changes.
Yes, I have read the README of the image in docker hub and no changes are needed in that scope.
@ofermend I think the workflow which publish the docker image, needs some changes. Can I do that?

**Comment by ofermend** [comment]
LGTM. I'll try locally and merge.
Thank you so much for this contribution!

**Review by ofermend — APPROVED** [review]
LGTM

**Comment by SaptarshiSarkar12** [comment]
@ofermend No, I have not tried with queries.json file. Sorry for that. But, I have found a technique to achieve this objective by defining volume in docker compose. I will make the change and push soon. 
Thank you for pointing out :smile:!

**Comment by SaptarshiSarkar12** [comment]
@ofermend One thing needs to be changed in the `ConfigurationContext.tsx` file. It is about the location of the queries.json file. I tried to use `fs` to check if the custom queries file is present - if it is, then, queries will be loaded from it.
But, I got this error :point_down: during the build time
```sh
Module not found: Error: Can't resolve 'fs' in '/src/contexts
```
@ofermend Can you please guide me how can I resolve this issue?
> [!NOTE]
> For docker, the queries.json if supplied, then, it gets located in `build/query` folder else there is nothing in query folder.

**Comment by ofermend** [comment]
Hey @SaptarshiSarkar12 - it may be that you are trying fs from the "client" side of the app (running on the user browser) which is not really possible. But I don't think you need to check this really, I would rather just make sure the file is there. You can for example login to the Docker container while it's running and just verify it's there.
Why do you need the location changed? Where are you trying to relocate queries.json to?

**Comment by SaptarshiSarkar12** [comment]
@ofermend The queries.json file if provided, then, it gets placed in `/usr/src/app/build/query` directory. But, if it is not provided, then, the current directory is placed in `/usr/src/app/build/query`. So, queries.json is not present in the `build` directory but in `build/query` directory. So, one way is to copy that out of that directory - which I tried just now and it unfortunately does not work :cry:. Other way is to edit the website code to fetch the `queries.json` file from the `query/` folder.
So, please tell me what you think can be done instead?

**Comment by SaptarshiSarkar12** [comment]
@ofermend Please tell me what should I do about the queries.json file? Would you agree with the queries.json file to be relocated in `query` directory? If yes, then, please tell me the areas where I should change the location - as far as I have seen, the change is required only in [**ConfigurationContext.tsx**](https://github.com/vectara/vectara-answer/blob/main/src/contexts/ConfigurationContext.tsx) file.

**Comment by ofermend** [comment]
sorry, can you please remind me why we have a need to change where queries.json is expected?

**Comment by SaptarshiSarkar12** [comment]
@ofermend I have already mentioned the reason in [this comment](https://github.com/vectara/vectara-answer/pull/105#issuecomment-1824794799). Let me know if you have any problem in understanding what I wanted to say.

**Comment by ofermend** [comment]
Oh I see. So if I understand correctly you are highlighting a separate issue regarding queries.json.
Typically I would recommend not to overload one PR with too many issues. This PR is meant to address issue 92, so let's focus on that. The issue you highlight (which I have not tested but think is absolutely correct) refers to if queries.json does not exist. 
We could solve it here too, but my recommendation would be to close this PR and do this separately. Does this make sense?

**Comment by SaptarshiSarkar12** [comment]
@ofermend The problem I am facing is to attach the **queries.json** file as a volume. That's why I was asking you to tell me if I can change the location of **queries.json** file - it must be relocated under **`build/query`** directory.

**Comment by SaptarshiSarkar12** [comment]
@ofermend Another method is to create a blank queries.json file whenever no correct config directory with queries.json file in it has been passed. By doing this, docker will not give any error that the host file does not exist.

**Comment by SaptarshiSarkar12** [comment]
@ofermend I think the second method is more effective and following that approach, brings less changes.

**Comment by ofermend** [comment]
Okay I got it now. So going back to the source - this [comment](https://github.com/vectara/vectara-answer/pull/105#pullrequestreview-1744998352), the issue is:
1. the "-v" flag in "docker run" does not work the same way with "docker compose"
2. Reading about it more here (https://docs.docker.com/storage/volumes/#use-a-volume-with-docker-compose) this is possible with docker-compose- do you know if it's possible to just replace the host path with an argument to docker-compose - like this
volumes:
   - ${HOST_PATH}:/path/in/container
Where HOST_PATH is defined as the local host just like today.

Honestly though one of my thoughts is: let's take a step back - do we really need docker-compose? What was the issue with just using "docker run" which does not have this issue?

**Comment by SaptarshiSarkar12** [comment]
@ofermend I have brought **docker compose** in the picture just to simplify the commands run :sweat_smile:.
So, what do you say? Should I remove it?

**Comment by ofermend** [comment]
Yes given the complication with volumes why not go back?

**Comment by SaptarshiSarkar12** [comment]
Okay, then, I'll revert my **docker compose** addition and make all the necessary changes.
Thank you for your constant guidance and immense help @ofermend, @cjcenizal and @irfanazam1 :smile:!

**Comment by SaptarshiSarkar12** [comment]
@ofermend Should the website port on the host system be 80 or 3000?

**Comment by SaptarshiSarkar12** [comment]
@ofermend I have made the requested changes. Please review this PR.

**Comment by ofermend** [comment]
> @ofermend Should the website port on the host system be 80 or 3000?

80

File: DOCKER.md
Comment (ofermend) [inline DOCKER.md:56]: Should the instructions here be reverted back to use "docker logs" and "docker stop"?

File: README.md
Comment (ofermend) [inline README.md:256]: Perhaps rewrite as "While `npm run start` runs the application with a local client..."

File: docker-compose.yml
Comment (ofermend) [inline docker-compose.yml:1]: Do we need coker-compose.yml anymore?

File: docker/run.sh
Comment (ofermend) [inline docker/run.sh:21]: This seems to still be using "docker compose" - let's switch back to just using "docker run" as discussed in the comment?

**Comment by SaptarshiSarkar12** [comment]
Sorry @ofermend. I think my changes have got removed while resolving the merge conflicts. I will make the necessary changes as soon as possible 😃.

File: README.md
Comment (SaptarshiSarkar12) [reply README.md:256]: Yeah, it can be done 👍. No problem 😁.

**Comment by SaptarshiSarkar12** [comment]
@ofermend I have made the requested changes. Please review this PR.

File: docker/run.sh
Comment (ofermend) [inline docker/run.sh:26]: Why did you take out the platform? This is important to keep in, as if you do not and run e.g. on Mac it generates the docker using ARM architecture which is not what we want (may create issue). Can we please put that back in?

Comment (SaptarshiSarkar12) [reply docker/run.sh:26]: I have removed the platform arg because it is already specified in the Dockerfile. I think that is enough. Let me know what you think 🤔.

Comment (ofermend) [reply docker/run.sh:26]: Ah, good point. Thank you - yes I think that should work.
Will test locally later today.

Comment (SaptarshiSarkar12) [reply docker/run.sh:26]: Okay :+1:.

**Comment by SaptarshiSarkar12** [comment]
@ofermend The builder stage does not take much time. I think you are talking about installing dependencies step.
I have tried two commands for the deps stage
- `npm install`
   
   [Screencast from 2023-11-28 08-53-50.webm](https://github.com/vectara/vectara-answer/assets/105960032/fb74599b-0390-4108-9e71-43ee41fea9c9)

- `npm ci`
   
   [Screencast from 2023-11-28 09-15-24.webm](https://github.com/vectara/vectara-answer/assets/105960032/ae52f0d4-1066-4a6f-ad03-c53596b19c23)

---
The `npm ci` is the faster of the two. It reduces the total docker image build time from **300 seconds** to just **200 seconds**. Moreover, the deps stage will be cached - so, time will further get reduced. The builder stage takes only **30 seconds** - which is not a much time.

**Comment by SaptarshiSarkar12** [comment]
@ofermend I have pushed the Dockerfile with `npm ci` command. Can you please check if that is still slow or not?

**Comment by SaptarshiSarkar12** [comment]
@ofermend You might have configured low ram and cpu usage in Docker desktop, which might be the reason for the long time taken for your build.

File: docker/run.sh
Comment (SaptarshiSarkar12) [reply docker/run.sh:26]: @ofermend Did this work?

**Comment by ofermend** [comment]
hmmm... I'm running with 12 cores and 16GB RAM; is that not enough?

**Comment by SaptarshiSarkar12** [comment]
@ofermend Can you please try again with the modified Dockerfile? I have replaced one command.

**Comment by ofermend** [comment]
I have cores=12 and 16GB RAM. With the update still does not finish. 
Could it just be an issue with Docker (see e.g. https://stackoverflow.com/questions/61629339/docker-build-takes-long-time-for-nodejs-application), when trying to build on the container?
Is there an issue to build locally and then copy ./build to the image?

**Comment by SaptarshiSarkar12** [comment]
@ofermend Okay, I see.
Can you please send a recording of your screen when you are running the docker build? I would like to inspect the problem.

**Comment by ofermend** [comment]
See this video.
This shows first 30 seconds, but I waited about 500-600 and still does not finish the build.
https://github.com/vectara/vectara-answer/assets/1823547/147a3296-6ce0-42df-a83f-bd441add3687

**Comment by SaptarshiSarkar12** [comment]
@ofermend Yeah, I saw. It is most probably a problem for your machine. It works fine in mine.
Still, can you try these commands :point_down:
```pwsh
docker rmi $(docker images -aq) --force
docker system prune
```
And then, can you please re-try the docker build and run?
I think it is a issue with your system. Also, can `--platform=linux/amd64` with `docker run`  command help in your case?

**Comment by ofermend** [comment]
Unfortunately none of these works. I truly think this is some kind of docker problem (and have seen a lot of others facing similar problems). What are your thoughts on just building locally and copying ./build? I realize it's not an ideal solution from a puristic DevOps perspective but worked so far, so I think my preference is to not spend too much more time on this and revert back to that option.

**Comment by SaptarshiSarkar12** [comment]
@ofermend Reverting will remove the cause for adding multi-stage docker build as for building, installing node modules is necessary which is already a part of the **deps** stage. So, no point in making this PR.
Rather, can you check [this StackOverflow answer](https://stackoverflow.com/a/76569834/19187952)? It has the exact problem you are facing. Might be it can help you.

**Comment by ofermend** [comment]
Yeah I get it. I know you've put so much work into this, but if doing the multi-stage build creates all these issues, then it's not worth the merge IMO. I look at the link you provided and tried what's suggested there as well as in others like it, but none of these helped. I wonder if we can ask the Docker community forum if there's a better solution?

**Comment by SaptarshiSarkar12** [comment]
@ofermend Inconsistency of systems create problems at times. I think that the Docker community could help us. They know these issues better.

**Comment by SaptarshiSarkar12** [comment]
@ofermend have you found any solution to the problem in your machine? Have you asked the Docker community?
I tried to join their slack channel but I couldn't because I think only docker members are allowed to join there.

**Comment by SaptarshiSarkar12** [comment]
@ofermend Did you get any solution?

**Comment by ofermend** [comment]
Nothing yet.
Here's the thread:
https://forums.docker.com/t/run-npm-run-build-takes-forever/138724.

**Comment by SaptarshiSarkar12** [comment]
@ofermend Okay. I see. Have you posted this question on StackOverflow? It might help posting there.

**Comment by SaptarshiSarkar12** [comment]
@ofermend Have you got any solution to the issue that you were facing?

**Comment by ofermend** [comment]
Sorry for late reply -  was on vacation
It seems to do with the difference in architectures between AMD and ARM. But the responder was not sure either - that was his hypothesis. I will check and let u know if I can confirm. Even if it is the case, I don't know what we can do - I don't think it's a good idea to prevent all mac users from being able to run this successfully.

**Comment by SaptarshiSarkar12** [comment]
@ofermend It's okay. I know that you might be on vacation because it's Winter and Christmas was there.
I think you should use **docker buildx** for building the docker image for any architecture you want. Let me give you an example.
I have built a [docker image](http://github.com/SaptarshiSarkar12/Drifty/pkgs/container/drifty-cli/) for my [Open-Source project Drifty](https://github.com/SaptarshiSarkar12/Drifty) using GitHub Action. You can see the **drifty-cli** docker image has mainly three architecture-wise images - **amd**, **arm64** and **arm/v7**. These images were generated by **docker-buildx** with the help of **QEMU**. You can see [this part of the workflow where I have defined the architectures for which the image needs to be built](https://github.com/SaptarshiSarkar12/Drifty/blob/master/.github/workflows/docker-publish.yml#L160).

**Comment by SaptarshiSarkar12** [comment]
@ofermend You can also try running [this docker image](https://github.com/SaptarshiSarkar12/Drifty/pkgs/container/drifty-cli) that has been built on linux for amd and arm and see if it works on your machine. If it works, then, I suggest you to use docker buildx for vectara.

**Comment by ofermend** [comment]
I see. So if I understand correctly - you are suggesting that the "deps" and "build" parts of the docker would be built by github actions, and pushed say to docker-hub. Then the final part of the build would pull from there to build "runner" locally and run. Is that the idea?

**Comment by SaptarshiSarkar12** [comment]
Yeah, but there's an issue. We can't expect every user to have their own Docker hub account and also we can't run GH actions locally. So, I was suggesting this idea for production use. @ofermend What do you think?

**Comment by ofermend** [comment]
Right now we have github action, as part of vectara-answer, to push any release of the docker image to Vectara's docker-hub. We can probably create a similar action for deps and build, and host them in the same place - I don't think pulling that image would then require a user authentication - am I wrong?
So in production or development, as long as you don't change the docker build/deps that much, you should be fine. And we can build an option to also redo those locally in dev mode. 
WDYT?

**Comment by SaptarshiSarkar12** [comment]
> Right now we have github action, as part of vectara-answer, to push any release of the docker image to Vectara's docker-hub.

@ofermend Then, I need to add **docker buildx** and **QEMU** in that workflow along with the changed docker commands.
> We can probably create a similar action for deps and build, and host them in the same place - I don't think pulling that image would then require a user authentication - am I wrong?

@ofermend We can't do this because the contributors will be using their modified version of the source code, which is certainly not going to match with the original src. So, that docker image won't help :smile:. I have an option. Docker buildx plugin can be used locally - I need to see it's commands and then, those commands can be included in the `run.sh` file. @ofermend What do you think?
In a nutshell, 
- For production, GH Action will handle the setting up of buildx, the building and pushing process.
- For development, docker buildx plugin install commands put in `run.sh` will handle the rest part (I hope so; I'll try and let you know if it is actually possible)
> [!NOTE]
> The dev build can take a little longer (from 2 to 4 minutes according to [this workflow run](https://github.com/SaptarshiSarkar12/Drifty/actions/runs/7355487224/usage)) due to emulation by QEMU and using Docker buildx.

**Comment by SaptarshiSarkar12** [comment]
@ofermend Can you please try rebuilding and running the docker image from my changes after running the below command?
```sh
export DOCKER_DEFAULT_PLATFORM=linux/amd64
```

**Comment by ofermend** [comment]
That didn't help.
However, I just upgrade Mac Docker Desktop to the latest version (just released - 4.26.1 (131620)), and now it works super quickly. I think they fixed issues with the speed in virtualization mode and now it works fine.
So I think let me retry this from scratch and ensure everything works okay, and perhaps we should be good as is. I'll just make a note to use Docker Desktop of this version or above for Mac users.

**Comment by ofermend** [comment]
Okay just checked locally on a clean pull of this branch with the new docker desktop and it works great.
Can you please merge from master the conflicting files above and we can prepare to merge?

**Comment by SaptarshiSarkar12** [comment]
> That didn't help. However, I just upgrade Mac Docker Desktop to the latest version (just released - 4.26.1 (131620)), and now it works super quickly. I think they fixed issues with the speed in virtualization mode and now it works fine. So I think let me retry this from scratch and ensure everything works okay, and perhaps we should be good as is. I'll just make a note to use Docker Desktop of this version or above for Mac users.

Okay. That's a pretty good news :tada:!

File: DOCKER.md
Comment (ofermend) [inline DOCKER.md:13]: Suggest adding here "For mac users, please ensure you Docker Desktop is version 4.26.1  or above"

**Comment by SaptarshiSarkar12** [comment]
> Okay just checked locally on a clean pull of this branch with the new docker desktop and it works great. Can you please merge from master the conflicting files above and we can prepare to merge?

That's awesome :tada: :partying_face:!
Yeah, I'll resolve the conflicts in some minutes.

File: DOCKER.md
Comment (SaptarshiSarkar12) [reply DOCKER.md:13]: Okay. I'll add and resolve this conversation.
Thank you for your review!

**Comment by SaptarshiSarkar12** [comment]
@ofermend I have added the Docker version requirement like this :point_down: 
image

**Review by ofermend — APPROVED** [review]
LGTM. Good job pushing through and being persistent. Glad we figured out the issue.

**Comment by SaptarshiSarkar12** [comment]
Thank you @ofermend for your support and cooperation :smile:!
I loved working on this issue and this project.

**Pull request merged**
